conmon-rs/server/src/oom_watcher.rs

```diff
@@ -0,0 +1,264 @@
+        memory_events_file_path: &Path,
+        last_counter: u64,
+    ) -> Result<(u64, bool)> {
+        let mut new_counter: u64 = 0;
+        let mut found_oom = false;
+        let fp = File::open(memory_events_file_path).await?;
+        let reader = BufReader::new(fp);
+        let mut lines = reader.lines();
+        while let Some(line) = lines.next_line().await? {
+            debug!("{}", line);
+            if let Some(counter) = line.strip_prefix("oom ") {
+                let counter = counter.to_string().parse::<u64>()?;
+                if counter != last_counter {
+                    new_counter = counter;
+                    found_oom = true;
+                    break;
+                }
+            }
+        }
+        Ok((new_counter, found_oom))
+    }
+
+    async fn write_oom_files(exit_paths: &[PathBuf]) -> Result<()> {
+        let paths = exit_paths.to_owned();
+        let tasks: Vec<_> = paths
+            .into_iter()
+            .map(|path| {
+                tokio::spawn(async move {
+                    let fp = File::create(path.join("oom")).await;
+                    drop(fp);
+                })
+            })
+            .collect();
+        for task in tasks {
+            task.await?;
+        }
+        Ok(())
+    }
+
+    async fn process_cgroup_subsystem_path(
+        pid: u32,
+        is_cgroupv2: bool,
+        subsystem: &str,
+    ) -> Result<PathBuf> {
+        if is_cgroupv2 {
+            Self::process_cgroup_subsystem_path_cgroup_v2(pid).await
+        } else {
+            Self::process_cgroup_subsystem_path_cgroup_v1(pid, subsystem).await
+        }
+    }
+
+    async fn process_cgroup_subsystem_path_cgroup_v1(pid: u32, subsystem: &str) -> Result<PathBuf> {
+        let cgroup_path = format!("/proc/{}/cgroup", pid);
+        debug!("{}: using cgroup path : {}", pid, cgroup_path);
+        let fp = File::open(cgroup_path).await?;
+        let reader = BufReader::new(fp);
+        let mut lines = reader.lines();
+        let re = Regex::new(r".*:(.*):/(.*)")?;
+        while let Some(line) = lines.next_line().await? {
+            debug!("{}: cgroup data : {}", pid, line);
+            if let Some(caps) = re.captures(&line) {
+                let system = caps[1].to_string();
+                let path = caps[2].to_string();
+                if system.contains(subsystem) || system.eq("") {
+                    return Ok(PathBuf::from(CGROUP_ROOT).join(subsystem).join(path));
+                }
+            }
+        }
+        Err(anyhow!("no path found"))
+    }
+
+    async fn process_cgroup_subsystem_path_cgroup_v2(pid: u32) -> Result<PathBuf> {
+        let fp = File::open(format!("/proc/{}/cgroup", pid)).await?;
+        let mut buffer = String::new();
+        let mut reader = BufReader::new(fp);
+        if reader.read_line(&mut buffer).await? == 0 {
+            Err(anyhow!("invalid cgroup"))
+        } else {
+            let re = Regex::new(r".*:.*:/(.*)")?;
```

Comment:
Do we wanna use lazy_static to precompile the regex ahead of time?

Reply:
fixed

Reply:
fixed.